src/hexkit/providers/akafka/provider/eventsub.py

```diff
@@ -181,76 +226,220 @@ async def construct(
-        self, *, consumer: KafkaConsumerCompatible, translator: EventSubscriberProtocol
+        self,
+        *,
+        consumer: KafkaConsumerCompatible,
+        translator: EventSubscriberProtocol,
+        config: KafkaConfig,
+        dlq_publisher: Optional[EventPublisherProtocol] = None,
     ):
         """Please do not call directly! Should be called by the `construct` method.
         Args:
             consumer:
-                hands over a started AIOKafkaProducer.
+                hands over a started AIOKafkaConsumer.
             translator (EventSubscriberProtocol):
                 The translator that translates between the protocol (mentioned in the
                 type annotation) and an application-specific port
                 (according to the triple hexagonal architecture).
+            dlq_publisher:
+                running instance of publishing provider that implements the
+                EventPublisherProtocol, such as KafkaEventPublisher. Can be None if
+                not using the dead letter queue. It is used to publish events to the DLQ.
+            config:
+                The KafkaConfig instance
         """
         self._consumer = consumer
         self._translator = translator
         self._types_whitelist = translator.types_of_interest
-
-    @staticmethod
-    def _get_event_label(event: ConsumerEvent) -> str:
-        """Get a label that identifies an event."""
-        return (
-            f"{event.topic} - {event.partition} - {event.key} - {event.offset}"
-            + " (topic-partition-offset)"
+        self._dlq_publisher = dlq_publisher
+        self._dlq_topic = config.kafka_dlq_topic
+        self._retry_topic = config.kafka_retry_topic
+        self._max_retries = config.kafka_max_retries
+        self._enable_dlq = config.kafka_enable_dlq
+        self._retry_backoff = config.kafka_retry_backoff
+
+    async def _publish_to_dlq(self, *, event: ExtractedEventInfo):
+        """Publish the event to the DLQ topic."""
+        logging.debug("About to publish an event to DLQ topic '%s'", self._dlq_topic)
+        await self._dlq_publisher.publish(  # type: ignore
+            payload=event.payload,
+            type_=event.type_,
+            topic=self._dlq_topic,
+            key=event.key,
+            headers={ORIGINAL_TOPIC_FIELD: event.topic},
         )
+        logging.info("Published event to DLQ topic '%s'", self._dlq_topic)
+
+    async def _retry_event(self, *, event: ExtractedEventInfo, retries_left: int):
+        """Retry the event until the maximum number of retries is reached.
+
+        If the retry fails, this method is called again with `retries_left` decremented.
+
+        Raises:
+        - `RetriesExhaustedError`: If all retries are exhausted without success.
+        - `ValueError`: If `retries_left` is invalid.
+        """
+        # Check if retries_left is valid.
+        if not 0 < retries_left <= self._max_retries:
+            error = ValueError(
```

Comment:
Maybe also raise a RetriesExhaustedError in this case since it probably should be dealt with in the same way.

Reply:
The `RetriesExhaustedError` represents a specific outcome -- that an event failed, was retried at least once, and failed for all of the allowed retries. The current logic should already prevent the function from being called with `retries_left` outside of the above criteria. If it somehow _does_ get called with an invalid value, `ValueError` is a much better fit to convey what happened than `RetriesExhaustedError`, even if the downstream handling is identical. Between logs and the traceback, I think this will provide the most clarity for diagnostics.

Reply:
Yes, it should never happen. I agree that it would not be really clean to reuse this error. Maybe raise a specific subclass of ValueError that you can catch to make it cleaner.

Reply:
Alright, I used a new error class in [27edd4a](https://github.com/ghga-de/hexkit/pull/123/commits/27edd4a45f1a0b25d7d66b8e8d008a1d576c7d1d), which also updates the comment mentioned above